.circleci/config.yml

```diff
@@ -1,173 +1,193 @@
-            rm cmd/promtail/promtail
-            make GOOS=windows BUILD_IN_CONTAINER=false promtail
+  publish/loki:
+    <<: *publish
+    environment:
+      APP: loki
 
-      - run:
-          name: Build Images
-          command: |
-            make images
+  # Loki
+  build/canary:
+    <<: *container
+    environment:
+      APP: loki-canary
 
-      - run:
-          name: Save Images
-          command: |
-            make save-images
+  publish/canary:
+    <<: *publish
+    environment:
+      APP: loki-canary
 
-      - save_cache:
-          key: v1-loki-{{ .Branch }}-{{ .Revision }}
-          paths:
-          - images/
+  # Promtail
+  build/promtail:
+    <<: *container
+    environment:
+      APP: promtail
 
-      - save_cache:
-          key: v1-loki-plugin-{{ .Branch }}-{{ .Revision }}
-          paths:
-          - cmd/docker-driver/docker-driver
+  publish/promtail:
+    <<: *publish
+    environment:
+      APP: promtail
 
-  publish:
+  # Docker driver
+  build/docker-driver:
     <<: *defaults
     steps:
       - checkout
       - setup_remote_docker
-
-      - restore_cache:
-          key: v1-loki-{{ .Branch }}-{{ .Revision }}
-      - restore_cache:
-          key: v1-loki-plugin-{{ .Branch }}-{{ .Revision }}
-
       - run:
-          name: Load Images
-          command: |
-            make load-images
+          name: docker-driver
+          command: make docker-driver
 
-      - run:
-          name: Push Images
-          command: |
-            if [ -n "$DOCKER_USER" ]; then
-              docker login -u "$DOCKER_USER" -p "$DOCKER_PASS" &&
-              make push-images
-            fi
-
-      - run:
-          name: Push Docker Plugin
-          command: |
-            if [ -n "$DOCKER_USER" ]; then
-              docker login -u "$DOCKER_USER" -p "$DOCKER_PASS" &&
-              make BUILD_IN_CONTAINER=false docker-driver-push
-            fi
-
-  publish-master:
+  publish/docker-driver:
     <<: *defaults
     steps:
       - checkout
       - setup_remote_docker
-
-      - restore_cache:
-          key: v1-loki-{{ .Branch }}-{{ .Revision }}
-      - restore_cache:
-          key: v1-loki-plugin-{{ .Branch }}-{{ .Revision }}
-
       - run:
-          name: Load Images
-          command: |
-            make load-images
-
+          name: login
+          command: docker login -u "$DOCKER_USER" -p "$DOCKER_PASS"
       - run:
-          name: Push Images
-          command: |
-            docker login -u "$DOCKER_USER" -p "$DOCKER_PASS" &&
-            make push-latest
-
-      - run:
-          name: Push Docker Plugin
-          command: |
-            if [ -n "$DOCKER_USER" ]; then
-              docker login -u "$DOCKER_USER" -p "$DOCKER_PASS" &&
-              PLUGIN_TAG=master make BUILD_IN_CONTAINER=false docker-driver-push && PLUGIN_TAG=latest make BUILD_IN_CONTAINER=false docker-driver-push
-            fi
+          name: docker-driver
+          command: make docker-driver-push
```

Comment:
There is just a bit of an inconsistency here we may want to fix, all the other images use the `*publish` which explicitly calls `make $APP-image` and then `make $APP-push`  however the docker-driver just calls `make docker-driver-push` and that make target has a pre-req on `docker-driver`.

I'm wondering if it's better to remove that pre-req and explicitly call both steps here, or make the other `$APP-push` targets have pre-reqs on the `$APP-image`.  I think the latter option of including the pre-req's is the better makefile style solution allowing you to just run `$APP-push` if you running things manually.

WDYT?

Reply:
Switched to make pre-reqs everywhere